### PR TITLE
CES-9801 - Helper functions to determine boot mode, BIOS or UEFI

### PR DESCRIPTION
**Pull request by digambar15**
Below tasks are implemented -
 

1.  Determine current boot mode
2.  Check is_boot_order_flexibly_programmable() method
3. Define dictionary to change the boot order from drac to ironic and vice-versa

### CODE REVIEW

File: boot_mode_helper.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    DRAC_BOOT_MODE_BIOS: boot_devices.BIOS,
+    DRAC_BOOT_MODE_UEFI: "uefi"
+}
+
+IRONIC_BOOT_MODES = dict((v, k) for (k, v) in DRAC_BOOT_MODES.items())
+
+
+class BootModeHelper(object):
+    path = os.path.basename(sys.argv[0])[0]
+    LOG = logging.getLogger(os.path.splitext(path)[0])
```
Comment (cdearborn) [inline boot_mode_helper.py]: The above 2 variables are not used, so they should be removed.

File: boot_mode_helper.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+import sys
```
Comment (cdearborn) [inline boot_mode_helper.py]: When the 2 unused variables below are removed, it looks like logging and sys will no longer be needed, and they should be removed from the imports.

File: boot_mode_helper.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/python
```
Comment (rpioso) [inline boot_mode_helper.py]: Please move this file to the `JetPack/src/pilot` directory.

File: boot_mode_helper.py (hunk 1 above)
Comment (digambar15) [reply boot_mode_helper.py]: Yeah, I will remove it

File: boot_mode_helper.py (hunk 2 above)
Comment (digambar15) [reply boot_mode_helper.py]: Done

File: boot_mode_helper.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        return 'SetBootOrderFqdd1' in bios_settings
+
+    @staticmethod
+    def determine_boot_mode(drac_client, bios_settings=None):
+        if not bios_settings:
+            bios_settings = drac_client.list_bios_settings(by_name=True)
+        if is_boot_order_flexibly_programmable(drac_client, bios_settings):
+            drac_boot_mode = bios_settings['BootMode'].current_value
+
+            return drac_boot_mode
```
[the remark this answers is not in the document]
Comment (digambar15) [reply boot_mode_helper.py]: I will discuss this with Richard in today's call.

[the remark this answers is not in the document]
Comment (digambar15) [reply boot_mode_helper.py]: @rpioso - can you reply on this ?

File: boot_mode_helper.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+DRAC_BOOT_MODE_UEFI = 'Uefi'
+
+# TODO(Update Ironic Boot Modes)
+"""For Rocky release, boot modes are defined in boot_modes.py file,
+When we upgrade to next version, we need to import ironic boot modes
+instead of [boot_devices.BIOS and uefi]."""
+DRAC_BOOT_MODES = {
+    DRAC_BOOT_MODE_BIOS: boot_devices.BIOS,
+    DRAC_BOOT_MODE_UEFI: "uefi"
+}
```
[the remark this answers is not in the document]
Comment (digambar15) [reply boot_mode_helper.py]: okay, will do.

File: boot_mode_helper.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+
+class BootModeHelper(object):
+    path = os.path.basename(sys.argv[0])[0]
+    LOG = logging.getLogger(os.path.splitext(path)[0])
+
+    @staticmethod
+    def is_boot_order_flexibly_programmable(drac_client, bios_settings=None):
+        if not bios_settings:
+            bios_settings = drac_client.list_bios_settings()
```
[the remark this answers is not in the document]
Comment (digambar15) [reply boot_mode_helper.py]: okay

File: boot_mode_helper.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        return 'SetBootOrderFqdd1' in bios_settings
+
+    @staticmethod
+    def determine_boot_mode(drac_client, bios_settings=None):
```
[the remark this answers is not in the document]
Comment (digambar15) [reply boot_mode_helper.py]: okay

File: boot_mode_helper.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (digambar15) [reply boot_mode_helper.py]: sure.

File: boot_mode_helper.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/python
+
+# Copyright (c) 2016-2018 Dell Inc. or its subsidiaries.
```
[the remark this answers is not in the document]
Comment (digambar15) [reply boot_mode_helper.py]: okay

File: boot_mode_helper.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+from ironic.common import boot_devices
+
+DRAC_BOOT_MODE_BIOS = 'Bios'
+DRAC_BOOT_MODE_UEFI = 'Uefi'
+
+# TODO(Update Ironic Boot Modes)
+"""For Rocky release, boot modes are defined in boot_modes.py file,
+When we upgrade to next version, we need to import ironic boot modes
+instead of [boot_devices.BIOS and uefi]."""
+DRAC_BOOT_MODES = {
```
[the remark this answers is not in the document]
Comment (digambar15) [reply boot_mode_helper.py]: Done

File: boot_mode_helper.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    DRAC_BOOT_MODE_BIOS: boot_devices.BIOS,
+    DRAC_BOOT_MODE_UEFI: "uefi"
+}
+
+IRONIC_BOOT_MODES = dict((v, k) for (k, v) in DRAC_BOOT_MODES.items())
```
[the remark this answers is not in the document]
Comment (digambar15) [reply boot_mode_helper.py]: Done

File: boot_mode_helper.py (hunk 3 above)
Comment (digambar15) [reply boot_mode_helper.py]: Yeah, Actually I got this after pushing the commit. will update

File: boot_mode_helper.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (digambar15) [reply boot_mode_helper.py]: okay.

File: src/pilot/boot_mode_helper.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+DRAC_BOOT_MODE_BIOS = 'Bios'
+DRAC_BOOT_MODE_UEFI = 'Uefi'
+
+DRAC_TO_IRONIC_BOOT_MODES = {
+    DRAC_BOOT_MODE_BIOS: deploy_utils.SUPPORTED_CAPABILITIES['boot_mode'][0],
+    DRAC_BOOT_MODE_UEFI: deploy_utils.SUPPORTED_CAPABILITIES['boot_mode'][1]
+}
+
+IRONIC_TO_DRAC_BOOT_MODES = \
```
Comment (rpioso) [inline src/pilot/boot_mode_helper.py]: Similar to the above comment, please change the name of the dictionary to

    IRONIC_BOOT_MODE_CAP_TO_DRAC_BOOT_MODE

A longer option is `IRONIC_BOOT_MODE_CAPABILITY_TO_DRAC_BOOT_MODE`.

Regardless of what is chosen, the names of the two (2) dictionaries should be consistent.

File: src/pilot/boot_mode_helper.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from ironic.drivers.modules import deploy_utils
+
+DRAC_BOOT_MODE_BIOS = 'Bios'
+DRAC_BOOT_MODE_UEFI = 'Uefi'
+
+DRAC_TO_IRONIC_BOOT_MODES = {
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/boot_mode_helper.py]: I will replace existing dictionary name with "DRAC_BOOT_MODE_TO_IRONIC_BOOT_MODE_CAP".
DRAC_BOOT_MODE_TO_IRONIC_BOOT_MODE_CAP - Looks good to me.

**Comment by digambar15** [comment]
All the changes are unit tested, below are the results -

is_boot_order_flexibly_programmable =  True
Get_boot_mode =  Bios
DRAC_BOOT_MODE_TO_IRONIC_BOOT_MODE_CAP =  {'Bios': 'bios', 'Uefi': 'uefi'}
IRONIC_BOOT_MODE_CAP_TO_DRAC_BOOT_MODE =  {'bios': 'Bios', 'uefi': 'Uefi'}

**Pull request merged**